BH-1197: import/export

matthewguywright opened the pull request:
## Description

Import/Export main feature branch.

matthewguywright committed:
BH-1196: add tools to submenu and placeholder page (#260)

* Added tools to submenu for ACM as well as the Tools page placeholder.

matthewguywright committed:
Merge branch 'main' into feature/tools

matthewguywright committed:
init commit buttons and text

matthewguywright committed:
Update Tools.jsx

matthewguywright committed:
latest

matthewguywright committed:
remove unneeded api call

anthonyburchell committed:
adds content model endpoint response

anthonyburchell committed:
adds higher level model endpoint

anthonyburchell committed:
revert composer and package changes

anthonyburchell committed:
add package-lock.json from main

matthewguywright committed:
remove unneeded api call

matthewguywright committed:
Merge branch 'feature/tools' of https://github.com/wpengine/atlas-content-modeler into feature/tools

matthewguywright committed:
Merge branch 'main' into feature/tools

matthewguywright committed:
export working for json

matthewguywright committed:
Update Tools.jsx

mindctrl commented on includes/settings/js/src/components/Tools.jsx:
```diff
@@ -0,0 +1,130 @@
+										"atlas-content-modeler"
+									)}
+								</p>
+								<button className="button button-primary link-button">
```
Looks like button styles are off? You see same?

What I see:
![image](https://user-images.githubusercontent.com/1254870/133494341-7cdc96b9-d144-4e55-bcd2-ec72184728bb.png)

Should look like:

![image](https://user-images.githubusercontent.com/1254870/133494375-18771496-8772-4747-b363-b40ed80c617d.png)

matthewguywright replied:
Updated.

mindctrl commented on includes/settings/js/src/components/Tools.jsx:
```diff
@@ -0,0 +1,130 @@
+					link.href = url;
+					link.click();
+
+					toast(
```
The toast design looks off to me. Here's what I see:

![image](https://user-images.githubusercontent.com/1254870/133494585-15d962a0-d3d9-4bba-850c-9d1b07cfe716.png)

matthewguywright replied:
This has been updated.

matthewguywright committed:
button isolated

matthewguywright committed:
updated export button

matthewguywright committed:
updated toast message design

matthewguywright committed:
Update ExportFileButton.jsx

matthewguywright committed:
Update Tools.jsx

matthewguywright committed:
Update ExportFileButton.jsx

matthewguywright committed:
updated title

matthewguywright committed:
update to allow file type

matthewguywright committed:
Update ExportFileButton.jsx

matthewguywright committed:
Update Tools.jsx

matthewguywright committed:
updated button color to match mock

matthewguywright committed:
button test

matthewguywright committed:
added file type ability for future options

mindctrl commented on includes/settings/js/src/components/Tools.jsx:
```diff
@@ -0,0 +1,85 @@
+							<div className="col-xs-12">
+								<h4>
+									{__(
+										"Import Model",
```
```suggestion
										"Import Models",
```

I know the mocks show singular, but it's more accurate to use plural.

mindctrl commented on includes/settings/js/src/components/Tools.jsx:
```diff
@@ -0,0 +1,85 @@
+							<div className="col-xs-12 mt-4">
+								<h4>
+									{__(
+										"Export Model",
```
```suggestion
										"Export Models",
```

Same as above.

mindctrl commented on includes/settings/js/src/components/Tools.jsx:
```diff
@@ -0,0 +1,85 @@
+								</h4>
+								<p className="help">
+									{__(
+										"Exporting a model will generate a .json document representing all of the existing models and fields.",
```
```suggestion
										"Exporting models will generate a .json document representing all of the existing models and fields.",
```

mindctrl commented on includes/settings/js/src/components/Tools.jsx:
```diff
@@ -0,0 +1,85 @@
+								</h4>
+								<p className="help">
+									{__(
+										"Select a .json file containing model/field definitions to import as a content model.",
```
```suggestion
										"Select the .json file containing model/field definitions to be imported.",
```

mindctrl commented on includes/settings/js/src/components/Tools.jsx:
```diff
@@ -0,0 +1,85 @@
+									)}
+								</p>
+								<ExportFileButton
+									fileTitle={`${getFormattedDateTime()}-model-export.json`}
```
I think we should include something here that identifies the file came from ACM. Something like `acm-models-export-DATETIME.json` or `atlas-content-modeler-models-export-DATETIME.json`

Also I think it helps to put the datetime at the end, so the filename "reads" more like a short sentence.

nickcernis replied:
I like `acm-models-export-DATETIME.json`.

mindctrl commented on includes/settings/scss/_buttons.scss:
```diff
@@ -47,6 +47,16 @@ button.tertiary {
+	&:focus,
+	&:hover {
+		background: lighten($color-blue-dark, 10%) !important;
+		border: 1px solid lighten($color-blue-dark, 10%) !important;
```
Can we make it so `!important` isn't required here? These are starting to creep more and more into our styles.

matthewguywright replied:
Updated

nickcernis commented on includes/settings/js/src/components/Tools.jsx:
```diff
@@ -0,0 +1,85 @@
+	 * Format filename for export
+	 * @returns {string}
+	 */
+	function getFormattedDateTime() {
+		var today = new Date();
+		var y = today.getFullYear();
+		var m = today.getMonth() + 1;
+		var d = today.getDate();
+		var h = today.getHours();
+		var mi = today.getMinutes();
+		var s = today.getSeconds();
+		return m + "-" + d + "-" + y + "-" + h + "-" + mi + "-" + s;
+	}
```
Worth considering a standard format like [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) with year-month-day?

For example:

```suggestion
	function getFormattedDateTime() {
		return new Date().toISOString().split('.')[0].replace(/[T:]/g, '-');
	}
```

Which gives:

```
2021-09-20-09-49-38
```

It's generally easier to sort/parse filenames in this format because sorting by name also sorts by date.

nickcernis commented:
@matthewguywright Welcome back! Just pinging in case you hand't seen the review comments above here. (Please feel free to ignore this if you had!)

matthewguywright committed:
Update includes/settings/js/src/components/Tools.jsx

Co-authored-by: John Parris <john.parris@wpengine.com>

matthewguywright committed:
Update includes/settings/js/src/components/Tools.jsx

Co-authored-by: John Parris <john.parris@wpengine.com>

matthewguywright committed:
Update includes/settings/js/src/components/Tools.jsx

Co-authored-by: John Parris <john.parris@wpengine.com>

matthewguywright committed:
Update includes/settings/js/src/components/Tools.jsx

Co-authored-by: John Parris <john.parris@wpengine.com>

matthewguywright committed:
Update includes/settings/js/src/components/Tools.jsx

Co-authored-by: John Parris <john.parris@wpengine.com>

matthewguywright committed:
Update includes/settings/js/src/components/Tools.jsx

Co-authored-by: Nick Cernis <nick@cern.is>

matthewguywright committed:
updated per PR

matthewguywright committed:
Merge branch 'feature/tools' of https://github.com/wpengine/atlas-content-modeler into feature/tools

matthewguywright committed:
Update Tools.jsx

matthewguywright committed:
Merge branch 'main' into feature/tools

nickcernis reviewed (dismissed):
Looking good — nice work!

I explored e2e tests with Codeception to verify the downloaded file using the [Filesystem module](https://codeception.com/docs/modules/Filesystem), but I couldn't figure out a way to set a download path for Chrome that worked on Circle and locally. I'm happy to skip this for now.

nickcernis committed:
Merge branch 'main' into feature/tools

matthewguywright committed:
Merge branch 'main' into feature/tools

matthewguywright committed:
Merge branch 'main' into feature/tools

matthewguywright committed:
Feature/mw bh 1198 import (#287)

Allows for import and export of ACM models

matthewguywright committed:
Merge branch 'main' into feature/tools

matthewguywright committed:
Merge branch 'feature/tools' of https://github.com/wpengine/atlas-content-modeler into feature/tools

nickcernis committed:
fix: clear file input field after upload attempts

So that the field's onChange event is still triggered if
the user selects the same file for a second upload attempt.

Props @mindctrl for suggesting we fix this.

nickcernis committed:
fix: re-fetch models before validating import data

So that changes outside of the current tab will be picked
up during validation. For example:

• I attempt an upload and see a “model exists” conflict.
• I delete all models via WP-CLI.
• My second upload attempt will now succeed without
having to refresh the import page.

Previously, a refresh was required because model state
was only set once on initial import page load.

nickcernis committed:
fix: only fetch models once per import

Instead of once per model in the import file.

ChrisWiegman reviewed (approved):
I finally got to review this this morning and could find no other issues. Well done!
